Use PostgreSQL instead of MongoDB

syuilo opened the pull request:
## Summary

PostgreSQLにするなどしています

Resolve #4534
Resolve #3510
Resolve #4203
Resolve #4018
Resolve #2497 
Resolve #2797
Resolve #3209
Resolve #3079
Resolve #2275
Resolve #2341
Resolve #4299
Resolve #3921
Resolve #2789
Resolve #4504
Resolve #1540
Resolve #4471
Resolve #4335
Close #1432
Close #1036
Close #1405
Close #4385 
Close #4377

* 削除系の処理が強化された
  * 例えば投稿を削除したらその投稿への返信やRenoteも削除され、その投稿への返信やRenoteに対する返信やRenoteも削除され... と連鎖し、関連するすべての情報がリンクして削除される
* IDを生成するときに投稿日時を考慮するようにした
  * これで古い投稿がタイムラインの先頭に出てくるなどの問題が修正される
* MongoDB固有機能のGridFSが使えなくなったことに伴い、オブジェクトストレージを使用しない場合、ドライブのファイルはサーバーのファイルシステム上に保存するように
* ユーザーリストに追加されているユーザー情報はユーザーリストに配列として埋め込むのではなく、新たに`UserListJoining`テーブルを作りそこで管理するように
  * これでユーザーが削除されたとき自動でリストからも削除されることになる
* ピン留めされている投稿情報はユーザーに配列として埋め込むのではなく、新たに`UserNotePinings`テーブルを作りそこで管理するように
  * これで投稿が削除されたとき自動でピン留めからも削除されることになる
* アンケートは別テーブルに分離
  * 個々の選択肢にIDを割り当てるのはやめて、選択肢配列のインデックスをID代わりにするように
* インスタンスのブロック情報はインスタンスレコードではなく`Meta`の`blockedHosts`プロパティに保存するように
  * 今まではいったん連合してからでないとブロックできなかった
* ユーザーの`publicKey`と`keypair`は別テーブルに分離
  * ほとんどの部分で必要とされないのでSELECTで無駄になる
* ホームのカスタマイズ情報はブラウザに保存するように

syuilo commented:
力尽きそう

rinsuki commented on src/server/api/endpoints/notes/sql.ts:
```diff
@@ -0,0 +1,124 @@
+	}
+};
+
+function sanitize(str: string): string {
```
自作のサニタイズ関数は漏れがあったときに怖いのでプレースホルダを使うべき

syuilo replied:
なるほど
プレースホルダについて調べたけどなんらかのライブラリを使うこと前提で実際にSQLがどうなるかがわからなかった

rinsuki replied:
だいたいは

```sql
SELECT * FROM hoge WHERE fuga=? AND piyo=?
```

を

```js
sql.prepare(sql_text).execute([1, 2])
```

で呼び出すと

```sql
SELECT * FROM hoge WHERE fuga=1 AND piyo=2
```

になる (他にも`{"hoge": 1}`を`:hoge`で参照できたり`$n`で配列の特定のところを参照できたりするやつもある)

syuilo replied:
db.prepare(sql_text).execute([1, 2]) というのはSQLじゃないから自前で実装するのかな

rinsuki replied:
だいたいの場合DBとやりとりしてくれるクライアントに入ってる

syuilo replied:
ふーむ、そうするとその関数はこのライブラリの外なので型定義が使えなくなっちゃいそう

rinsuki replied:
SQL文を作るときは[string, any[]]を返して、あとはラッパーで頑張るとよさそう

syuilo replied:
他のライブラリ見てみたけど自前でエスケープ処理書いてた(しかもめっちゃシンプルだった)
https://github.com/sequelize/sequelize/blob/master/lib/sql-string.js#L71

buoyantair replied:
I wish Github had a translate button ;-;

rinsuki commented:
SQL文ビルダー、便利そうなのでライブラリに切り出してnpmで公開してくれると他のプロジェクトでも使えてありがたい

syuilo commented:
JSON型の特定のキーの値をインクリメントする方法がわからない

syuilo commented:
起動するようにはなった

acid-chicken commented on .config/example.yml:
```diff
@@ -43,13 +43,42 @@ url: https://example.tld/
+################################################################
+#[ PostgreSQL configuration ]###################################
+
+db:
```
v11以降では

```yaml
db:
  type: postgresql
```

みたいな感じにして、v10は

```yaml
db:
  type: mongodb
```

に変更する（互換性は維持）とかにした方が綺麗そう。

syuilo commented:
アカウント作れるようになった

syuilo committed:
wip

syuilo commented:
投稿できるようになった

syuilo committed:
Update note.ts

syuilo committed:
Update timeline.ts

syuilo committed:
Update core.ts

syuilo committed:
wip

syuilo committed:
Update generate-visibility-query.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update global-timeline.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update vote.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update create.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update files.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update CONTRIBUTING.md

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo commented:
少し動くようにはなってきました
暇な方はお試しいただけると

syuilo committed:
wip

syuilo committed:
Update read-notification.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update cancel.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update show.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update gen-id.ts

syuilo committed:
Update create.ts

syuilo commented:
IDの生成方法を変更しました
Snowflake ID のように時間を考慮するようにしたので、リモートから来た古い投稿がタイムラインの先頭に表示されるなどの問題が解決されます

syuilo committed:
Update id.ts

syuilo committed:
wip

mei23 commented:
> Snowflake ID のように時間を考慮するようにしたので、リモートから来た古い投稿がタイムラインの先頭に表示されるなどの問題が解決されます

おそらくIDを生成日時基準で生成するということと思われますが
到着分をTL系API取得した場合はIDでソートすることで生成順で取得できますが
新規に到着した投稿がストリーミングで流れて来る場合は、
IDとかでソートする余地がなく常に到着順で流れてくるので
おそらく完全にはなおらないと思います。(なおせないと思います)

また、IDを生成日時順にしてしまうとTLが
到着順のストリーミング＋生成順のAPIから取得分 の混合になり
リロードすると順番が変わるという状態になるので
一長一短があると思います。

syuilo commented:
> 新規に到着した投稿がストリーミングで流れて来る場合は、

そこは改善予定で、投稿から一定時間経ったものはストリーミングに流さないようにするか、もしくは間接的にその投稿を取得した場合(例えばリモートの投稿の返信など)はストリーミングに流さない などで解決すると思ってます

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

mei23 commented on src/misc/gen-id.ts:
```diff
@@ -0,0 +1,38 @@
+function getRandom() {
+	let str = '';
+
+	for (let i = 0; i < 2; i++) {
```
ランダム部分10bit相当はなんとなく短い感じがしません？
完全ランダムでbit数少ないと衝突しそうな気がします

ランダム部分は
UUID: 122bit
ULID: 80bit
Twitter: マシンID (10bit) + カウンタ (12bit)
とかみたいです

syuilo replied:
同一ミリ秒以内に複数の投稿がなされる可能性は低いと思ったのでそのくらいで良いと思いました
でももうちょっと増やしても良いかもしれないですね

syuilo replied:
UUIDは「全世界で一意」を目指しているのでその長さだと思いますがMisskeyの場合は「インスタンス内で一意」であれば良いというのもあります

mei23 replied:
完全ランダムの実装の場合80bit程度は確保されてるみたいなので
何か裏付けられた理屈ががありそうで気になってます

mei23 replied:
あと、Base32とBase36の表現能力あまり変わらなそうなので
わかりやすさを考えると32でもいいかもしれません

Base32 2文字: 10bit
Base36 2文字: 10.3bit

Base32 4文字: 20bit
Base36 4文字: 20.7bit

Base32 8文字: 40bit
Base36 8文字: 41.3bit

acid-chicken replied:
めいめいの言う通り、この長さでた連続稼働していると普通に衝突する機会がありそうです。短くしたいのであればせめて衝突検知だけはしたほうがいい気がします。

syuilo replied:
衝突したらデータベースに挿入できずエラーになるのでユーザーがリトライすればOKそうです

mei23 replied:
衝突するのがユーザー起因の操作とは限らないので、リトライではちょっと厳しそうです。

mei23 replied:
やっぱり16文字80bitとかに…

mei23 replied:
@syuilo ランダム部分どれかに変えません？
2文字 / 10bit / 1296通り
↓
4文字 / 20bit / 167万通り
8文字 / 40bit
16文字 / 80bit

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

khws4v1 committed:
Docker: Update files about Docker (#4599)

* Docker: Use cache if files used by `yarn install` was not updated

This patch reduces the number of times to installing node_modules.
For example, `yarn install` step will be skipped when only ".config/default.yml" is updated.

* Docker: Migrate MongoDB to Postgresql

Misskey uses Postgresql as a database instead of Mongodb since version 11.

* Docker: Uncomment about data persistence

This patch will save a lot of databases.

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update activitypub.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update logs.ts

syuilo committed:
wip

syuilo committed:
Update drive-file.ts

syuilo committed:
Update register.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update mentions.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update recommendation.ts

syuilo committed:
wip

syuilo committed:
Update index.ts

syuilo committed:
wip

syuilo committed:
Update recommendation.ts

khws4v1 committed:
Doc: Update docker.ja.md and docker.en.md (#1) (#4608)

Update how to set up misskey.

syuilo committed:
wip

syuilo committed:
:v:

syuilo committed:
wip

syuilo committed:
Update note.ts

syuilo committed:
Update postgre.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update add-file.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Clean up

syuilo committed:
Update logs.ts

syuilo committed:
wip

syuilo committed:
:pizza:

syuilo committed:
wip

syuilo committed:
Ad notes

syuilo committed:
wip

syuilo committed:
Update api-visibility.ts

mei23 commented on src/services/drive/add-file.ts:
```diff
@@ -91,58 +85,51 @@ async function save(path: string, name: string, type: string, hash: string, size
-			contentType: type
-		});
-		//#endregion
+		file.url = url;
```
ここに`file.storedInternal = false;`がないと

`ERR  1	[api]	QueryFailedError: null value in column "storedInternal" violates not-null constraint`
になります

syuilo replied:
修正しました🙏

mei23 commented on src/models/repositories/note.ts:
```diff
@@ -0,0 +1,210 @@
+			userId: note.userId,
+			user: Users.pack(note.user || note.userId, meId),
+			text: text,
+			visibility: note.visibility,
```
`cw: note.cw,`がぬけてます

syuilo replied:
修正しました🙏

syuilo committed:
Update note.ts

syuilo committed:
Update add-file.ts

syuilo committed:
tests

syuilo committed:
tests

syuilo committed:
Update postgre.ts

syuilo committed:
Update utils.ts

syuilo commented:
https://kanonji.hatenadiary.com/entry/20110714/1310611877

これどうしよう
ユーザー名とホストでユニークなインデックス作っても、ローカル内ならいくらでも同じユーザー名でアカウント挿入できちゃう(ホストが`null`だとユニークにならないため)

syuilo commented:
新たにローカルユーザー専用のユーザー名カラムを作ってそれにユニーク制約付けるとか...？

syuilo commented:
まあ、なんにせよアカウント削除した後に同じユーザー名でアカウント作成できちゃう問題を防ぐために、ユーザーテーブルとは別にユーザー名テーブルも必要と思われるのでそっちでユニークチェックすれば良さそう

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Refactor

syuilo committed:
wip

syuilo committed:
Refactor

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
Update show-users.ts

syuilo committed:
Update update-instance.ts

syuilo committed:
wip

syuilo committed:
Update feed.ts

syuilo committed:
Update outbox.ts

syuilo committed:
Update outbox.ts

syuilo commented:
ほとんど終わった

syuilo committed:
Update user.ts

syuilo committed:
wip

syuilo committed:
Update list.ts

syuilo committed:
Update update-hashtag.ts

syuilo committed:
wip

syuilo committed:
Update update-hashtag.ts

syuilo committed:
Refactor

syuilo committed:
Update update.ts

syuilo committed:
wip

syuilo committed:
wip

syuilo committed:
:v:

syuilo committed:
clean up

syuilo committed:
docs

syuilo committed:
Update push.ts

acid-chicken commented:
> * 削除系の処理が強化された
>   * 例えば投稿を削除したらその投稿への返信やRenoteも削除され、その投稿への返信やRenoteに対する返信やRenoteも削除され... と連鎖し、関連するすべての情報がリンクして削除される

投稿の返信までもが削除されるのは要検討な気がします。Renote の場合は元の投稿をそのまま Announce するのでさほど問題ではありませんが、返信はあくまでも元の投稿にリンクされた別オブジェクトであるため、カンバセーションの管理権が返信元のユーザーに一元管理されてしまうのはあまり良いとは思えません。（ある投稿を引用して任意の投稿を作成する場合、本来は投稿の引用機能を利用するのが望ましいと想定されているでしょう。しかし、これは ActivityPub 上では非正規の方法で配信されているため、Fediverse 全体で見た場合は実際に引用したりする場合には返信の方が望ましい手段となってしまっているのが現状です。完全な機能の分離を図りたいのであれば、この問題を先に解決すべきでしょう
。）また、これらの実装は Mastodon などを含めた他の ActivityPub 実装とも頻繁に不整合が発生すると思われます。

syuilo committed:
wip

syuilo committed:
Update api.ts

syuilo committed:
wip

syuilo commented:
整合性が保てないので返信を残しておくのは難しそう
replyIdをNULLにしたりはできるけど、返信なのに返信ではない普通の投稿のように見えるようになっておかしくなりそう

mei23 commented:
> 返信なのに返信ではない普通の投稿のように見えるようになっておかしくなりそう

これくらいはしょうがないかも

tamaina commented:
replyIdをnullとundefinedで区別して（Twitterのように）もともと返信先があった場所に「投稿は削除されました」と表示するのはどうでしょう

syuilo commented:
PosgreSQLにはundefinedという概念が無いので無理そう

mei23 commented:
> replyIdをnullとundefinedで区別して

SQLにundefinedなどないのでは

tamaina commented:
（そういえばMongoDBではなかった）  
んーまあ、区別できればいいんです()

syuilo commented:
あとリプライ先のユーザーIDなどを非正規化していて、それらはリプライ先が削除されたとしてもNULLになったりしないので何らかの検索処理などで不具合が発生したりしそう
それらのことを考えるといっそリプライ自体も削除した方が良いのではないかという考えです
私もできれば現在のようにリプライ自体は残して、リプライ先を（削除された投稿）のようにしたいですが、手間をかけてまでそれを実現しようとしなくてもいいのかなと
一応手間をかけてそれを実現する方法としては、replyIdとは別にisReplyのようなプロパティを別途保存しておくとか...

syuilo committed:
:v:

syuilo committed:
Update make-pagination-query.ts

syuilo committed:
:v:

syuilo committed:
Delete hashtags.ts

syuilo committed:
Update instances.ts

syuilo committed:
Update instances.ts

syuilo committed:
Update create.ts

syuilo committed:
Update search.ts

syuilo committed:
Update reversi-game.ts

syuilo committed:
Update signup.ts

syuilo committed:
Update user.ts

syuilo committed:
id

syuilo commented:
IDの生成方式を選択できるようにした

syuilo committed:
Update example.yml

syuilo committed:
:art:

syuilo committed:
objectid

syuilo commented:
ObjectID生成を実装

syuilo committed:
fix

syuilo committed:
reversi

syuilo committed:
reversi

syuilo committed:
Fix bug of chart engine

syuilo committed:
Add test of chart engine

syuilo committed:
Improve test

syuilo committed:
Better testing

syuilo committed:
Improve chart engine

syuilo committed:
Refactor

syuilo committed:
Add test of chart engine

syuilo committed:
Refactor

syuilo committed:
Add chart test

syuilo committed:
Fix bug

syuilo commented:
チャートエンジンのテストを追加

syuilo committed:
コミットし忘れ

syuilo committed:
Refactoring

syuilo committed:
:v:

syuilo committed:
Add tests

mei23 commented:
アンケート作ってプロセスリスタートするとタイムラインとか壊れちゃう

syuilo committed:
Add test

syuilo committed:
Extarct note tests

syuilo committed:
Refactor

syuilo committed:
存在しないユーザーにメンションできなくなっていた問題を修正

syuilo committed:
Fix bug

syuilo committed:
Update update-meta.ts

syuilo committed:
Fix bug

syuilo committed:
Update mention.vue

syuilo committed:
Fix bug

syuilo committed:
Update meta.ts

syuilo commented:
* 存在しないユーザーにメンションできなくなっていた問題を修正
  * テスト追加
* インスタンス設定が更新できない問題を修正
* リモートのユーザーを取得時に non-nullなんとかが発生する問題を修正

syuilo committed:
Update CONTRIBUTING.md

syuilo committed:
Fix bug

syuilo committed:
Fix bug

syuilo committed:
Fix bug

syuilo commented:
いくつかのバグを修正

syuilo committed:
Clean up

syuilo committed:
Clean up

syuilo committed:
Update notification.ts

syuilo committed:
Clean up

syuilo committed:
Add mute tests

syuilo commented:
ミュートに関するテストを追加

syuilo committed:
Add test

syuilo committed:
Refactor

syuilo committed:
Add test

syuilo committed:
Fix test

syuilo committed:
Refactor

syuilo committed:
Refactor

syuilo committed:
Add tests

syuilo committed:
Update utils.ts

syuilo committed:
Update utils.ts

syuilo committed:
Fix test

syuilo committed:
Update package.json

syuilo committed:
Merge branch 'develop' into pg

syuilo committed:
Update update.ts

syuilo committed:
Update manifest.ts

syuilo committed:
Fix bug

syuilo committed:
Fix bug

syuilo committed:
Merge branch 'develop' into pg

syuilo committed:
Add test

syuilo committed:
:art:

syuilo committed:
Update endpoint permissions

syuilo committed:
Updaye permisison

syuilo committed:
Update person.ts

#4299

syuilo committed:
データベースと同期しないように

syuilo commented:
プロセス起動毎にデータベースとスキーマ同期しないようにして配列カラムが初期化される問題を回避できました
Misskeyのインストール時にデータベースのスキーマを初期化するステップが追加されています

syuilo committed:
Fix bug

syuilo committed:
Fix bug

syuilo commented:
いくつかのバグ修正

syuilo committed:
Update reversi-game.ts

syuilo commented:
マージして大丈夫そうかしら

tamaina committed:
Use a feature of Node v11.7.0 to extract a public key (#4644)

* wip

* wip

* :v:

syuilo committed:
Refactoring

#1540

tamaina commented:
https://github.com/syuilo/misskey/blob/pg/cli/mark-admin.js はどうしますか？

syuilo committed:
test

syuilo committed:
test

syuilo committed:
test

syuilo committed:
test

syuilo committed:
test

syuilo committed:
test

syuilo committed:
test

syuilo committed:
Fix bug

syuilo committed:
Fix test

syuilo committed:
:sushi:

syuilo commented:
投稿作成の処理が複雑になっているので各種ストリームに流す処理を削除
その代わり一つの投稿ストリームにだけ流すようにして、各種ストリームのチャンネルがそれを受け取って各々で必要な投稿をフィルタリングするようにした(wip)

syuilo committed:
wip

syuilo committed:
#4471

syuilo committed:
Add test for #4335

syuilo committed:
Refactor

syuilo committed:
Fix test

syuilo committed:
Add tests

syuilo committed:
:clock4:

syuilo committed:
Fix bug

syuilo committed:
Add test

syuilo commented:
ストリーミングのテスト追加

syuilo committed:
Add test

syuilo committed:
rename

syuilo committed:
Fix bug

poVoq commented:
Any chance for MySQL support as well?
